src/audio_input/audio.c

```diff
@@ -0,0 +1,205 @@
+      return false;
+   }
+   rsLog("Opened audio device %s", devname);
+   return true;
+}
+
+static void calcAudioParams(RSAudio *audio) {
+   if (audio->data.data != NULL) {
+      return;
+   }
+   int size1s = audio->ospec.channels * audio->ospec.freq * (int)sizeof(uint16_t);
+   int sizeTotal = size1s * audio->duration;
+   int numOfCallbacks =
+       (sizeTotal + (SAMPLES_PER_CALLBACK - 1)) / SAMPLES_PER_CALLBACK; // round up div
+   sizeTotal = numOfCallbacks * SAMPLES_PER_CALLBACK;
+   audio->sizeBatch = (size_t)(size1s / audio->framerate);
+   rsCircleStaticCreate(&audio->data, (size_t)sizeTotal);
+   audio->deviceAddTime = 0;
+   audio->deviceRemoveTime = 0;
+}
+
+static void deviceReconnect(RSAudio *audio, const char *devname) {
+   SDL_CloseAudioDevice(audio->deviceId);
+   /*
+    * set the audio buffor index to 0. Effectively erasing all
+    * the collected samples so far. This ensures that we'll be
+    * back in sync after the audio setup is complete.
+    * Once the frame pace and audio timestamps are implemented
+    * it should be calculated how long the audio setup took
+    * and move the index accordingly
+    */
+   audio->data.index = 0;
+   probeDevices(audio);
+   char *newname = getDeviceName(devname);
+   bool ok = openDevice(audio, newname);
+   rsMemoryDestroy(newname);
+   if (!ok) {
+      if (newname != NULL && audio->systemFallback) {
+         ok = openDevice(audio, NULL);
+      }
+      if (!ok) {
+         rsLog("Couldn't open any audio device. You will have no sound");
+         return;
+      }
+   }
+   calcAudioParams(audio);
+   SDL_PauseAudioDevice(audio->deviceId, 0);
+}
+
+void rsAudioCreate(RSAudio *audio, const RSConfig *config) {
+   rsMemoryClear(audio, sizeof(*audio));
+   if (SDL_Init(SDL_INIT_AUDIO | SDL_INIT_EVENTS) < 0) {
+      rsError("SDL2 could not initialize! SDL Error: %s", SDL_GetError());
+   }
+   SDL_EventState(SDL_AUDIODEVICEADDED, SDL_ENABLE);
+   SDL_EventState(SDL_AUDIODEVICEREMOVED, SDL_ENABLE);
+   if (pthread_spin_init(&audio->sampleGetLock, PTHREAD_PROCESS_PRIVATE)) {
+      rsError("pthread_spin_init failed");
+   }
+   audio->systemFallback = (config->audioSystemFallback != 0);
+   audio->duration = config->duration;
+   audio->framerate = config->framerate;
+
+   audio->ispec.freq = (int)config->audioSamplerate;
+   audio->ispec.format = AUDIO_S16LSB;
+   audio->ispec.channels = (Uint8)config->audioChannels;
+   audio->ispec.samples = SAMPLES_PER_CALLBACK; // must be power of 2
+   audio->ispec.callback = audioCallback;
+   audio->ispec.userdata = audio;
+
+   probeDevices(audio);
+   char *deviceName = getDeviceName(config->audioDeviceName);
+   if (!openDevice(audio, deviceName)) {
+      bool ok = false;
+      if (deviceName != NULL && audio->systemFallback) {
+         ok = openDevice(audio, NULL);
+      }
+      if (!ok) {
+         rsLog("Couldn't open any audio device. You will have no sound");
+      }
+   }
```

Comment:
Duplicate code of `deviceReconnect` above.

Reply:
done